drivers/i2c/i2c_mchp_mss.c

```diff
@@ -0,0 +1,399 @@
+#define STATUS_S_GENERAL_CALL_RX_DATA_ACKED  (0x90)
+#define STATUS_S_GENERAL_CALL_RX_DATA_NACKED (0x98)
+#define STATUS_S_RX_STOP                     (0xA0)
+#define STATUS_S_SLAR_ACKED                  (0xA8)
+#define STATUS_S_ARB_LOST_SLAR_ACKED         (0xB0)
+#define STATUS_S_TX_DATA_ACK                 (0xB8)
+#define STATUS_S_TX_DATA_NACK                (0xC0)
+#define STATUS_LAST_DATA_ACK                 (0xC8)
+
+#define PCLK_DIV_960 (CTRL_CR2)
+#define PCLK_DIV_256 (0)
+#define PCLK_DIV_224 (CTRL_CR0)
+#define PCLK_DIV_192 (CTRL_CR1)
+#define PCLK_DIV_160 (CTRL_CR0 | CTRL_CR1)
+#define PCLK_DIV_120 (CTRL_CR0 | CTRL_CR2)
+#define PCLK_DIV_60  (CTRL_CR1 | CTRL_CR2)
+#define BCLK_DIV_8   (CTRL_CR0 | CTRL_CR1 | CTRL_CR2)
+#define CLK_MASK     (CTRL_CR0 | CTRL_CR1 | CTRL_CR2)
+
+/* -- Transactions types -- */
+#define NO_TRANSACTION                     (0x00)
+#define CONTROLLER_WRITE_TRANSACTION       (0x01)
+#define CONTROLLER_READ_TRANSACTION        (0x02)
+#define CONTROLLER_RANDOM_READ_TRANSACTION (0x03)
+#define WRITE_TARGET_TRANSACTION           (0x04)
+#define READ_TARGET_TRANSACTION            (0x05)
+
+#define MSS_I2C_RELEASE_BUS (0x00)
+#define MSS_I2C_HOLD_BUS    (0x01)
+#define TARGET_ADDR_SHIFT	(0x01)
+
+typedef void (*mss_i2c_cfg_func_t)(void);
+
+enum i2c_status {
+	MSS_I2C_SUCCESS = 0u,
+	MSS_I2C_IN_PROGRESS,
+	MSS_I2C_FAILED,
+	MSS_I2C_TIMED_OUT
+};
+
+struct mss_i2c_config {
+	uint32_t clock_freq;
+	uintptr_t i2c_base_addr;
+	uint32_t i2c_irq_base;
+	void (*irq_config_func)(void);
+};
+
+struct mss_i2c_data {
+	uint8_t ser_address;
+	uint8_t target_addr;
+	uint8_t options;
+	uint8_t transaction;
+	const uint8_t *controller_tx_buffer;
+	uint16_t controller_tx_size;
+	uint16_t controller_tx_idx;
+	uint8_t dir;
+	uint8_t *controller_rx_buffer;
+	uint16_t controller_rx_size;
+	uint16_t controller_rx_idx;
+	atomic_t controller_status;
+	uint32_t controller_timeout_ms;
+	const uint8_t *target_tx_buffer;
+	uint16_t target_tx_size;
+	uint16_t target_tx_idx;
+	uint8_t *target_rx_buffer;
+	uint16_t target_rx_size;
+	uint16_t target_rx_idx;
+	atomic_t target_status;
+	uint8_t target_mem_offset_length;
+	uint8_t is_target_enabled;
+	uint8_t bus_status;
+	uint8_t is_transaction_pending;
+	uint8_t pending_transaction;
+
+	sys_slist_t cb;
+};
+
+
+static int mss_i2c_configure(const struct device *dev, uint32_t dev_config_raw)
+{
+	const struct mss_i2c_config *cfg = dev->config;
+
+	uint8_t ctrl = sys_read8(cfg->i2c_base_addr + CORE_I2C_CTRL);
+
+	switch (I2C_SPEED_GET(dev_config_raw)) {
+	case I2C_SPEED_STANDARD:
+		sys_write8((ctrl | PCLK_DIV_960), cfg->i2c_base_addr + CORE_I2C_CTRL);
+		break;
+	case I2C_SPEED_FAST:
+		sys_write8((ctrl | PCLK_DIV_256), cfg->i2c_base_addr + CORE_I2C_CTRL);
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static int mss_wait_complete(const struct device *dev)
+{
+	volatile struct mss_i2c_data *data = dev->data;
+	enum i2c_status i2c_status = data->controller_status;
```

Comment:
`controller_status` is now an `atomic_t` so the atomic helpers must be used when accessing / setting it. See `include/zephyr/sys/atomic.h`. This comment is valid for all the occurrences in this file.

Reply:
@carlocaione We are working towards this, hopefully we will finish up our testing locally today.